crates/walrus-service/src/event/event_processor/coordination.rs

```diff
@@ -0,0 +1,243 @@
+/// catchup manager during runtime catchup operations using message-passing.
+/// The coordination works in phases:
+/// 1. Download phase: Event blobs are downloaded in parallel with checkpoint tailing
+/// 2. Processing phase: Send StopCheckpointTailing message, process event blobs
+/// 3. Resume phase: Send RestartCheckpointTailing message (picks up from new latest checkpoint)
+#[derive(Debug)]
+pub struct CatchupCoordinationState {
+    /// Channel for sending coordination messages to checkpoint tailing task
+    pub coordination_tx: mpsc::UnboundedSender<CoordinationMessage>,
+    /// Whether catchup is currently active
+    pub catchup_active: Arc<std::sync::atomic::AtomicBool>,
+    /// Timestamp when catchup was last started
+    pub last_catchup_start: Arc<Mutex<Option<Instant>>>,
+    /// Notifier signaled when checkpoint tailing has fully stopped
+    tailing_stopped_notify: Arc<Notify>,
+    /// Flag indicating whether checkpoint tailing is currently stopped
+    is_tailing_stopped: Arc<std::sync::atomic::AtomicBool>,
+}
+
+impl CatchupCoordinationState {
+    /// Creates a new coordination state instance and returns it with a receiver.
+    pub fn new() -> (Self, mpsc::UnboundedReceiver<CoordinationMessage>) {
+        let (coordination_tx, coordination_rx) = mpsc::unbounded_channel();
+
+        (
+            Self {
+                coordination_tx,
+                catchup_active: Arc::new(std::sync::atomic::AtomicBool::new(false)),
+                last_catchup_start: Arc::new(Mutex::new(None)),
+                tailing_stopped_notify: Arc::new(Notify::new()),
+                is_tailing_stopped: Arc::new(std::sync::atomic::AtomicBool::new(false)),
+            },
+            coordination_rx,
+        )
+    }
+
+    /// Mark that checkpoint tailing has started.
+    pub fn mark_tailing_started(&self) {
+        self.is_tailing_stopped
+            .store(false, std::sync::atomic::Ordering::Release);
+    }
+
+    /// Notify that checkpoint tailing has fully stopped and wake any waiters.
+    pub fn notify_tailing_stopped(&self) {
+        self.is_tailing_stopped
+            .store(true, std::sync::atomic::Ordering::Release);
+        self.tailing_stopped_notify.notify_one();
+    }
+
+    /// Wait until checkpoint tailing is stopped, with timeout.
+    /// Returns true if stopped, false on timeout.
+    pub async fn wait_for_tailing_stopped(&self, timeout: std::time::Duration) -> bool {
+        if self
+            .is_tailing_stopped
+            .load(std::sync::atomic::Ordering::Acquire)
+        {
+            return true;
+        }
+        match tokio::time::timeout(timeout, self.tailing_stopped_notify.notified()).await {
+            Ok(()) => true,
+            Err(_) => false,
+        }
+    }
+
+    /// Start the download phase of catchup.
+    /// During this phase, event blobs are downloaded in parallel with checkpoint tailing.
+    pub async fn start_catchup_download_phase(&self) {
```

Comment:
I think this is not called anywhere other than in tests.

Looks like the purpose of setting catchup_active to true is to prevent multiple catch up tasks are initiated, is that the case? Right now, catchup_active is also never set.

Reply:
Yes, this is fixed now and is being invoked as expected, thanks!